src/main/java/org/ethereum/beacon/discovery/storage/KBuckets.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.ethereum.beacon.discovery.storage;
+
+import java.time.Clock;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.apache.tuweni.bytes.Bytes;
+import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.util.Functions;
+
+public class KBuckets {
+  public static final int MAXIMUM_BUCKET = 256;
+
+  private final LocalNodeRecordStore localNodeRecordStore;
+  private final Bytes homeNodeId;
+  private final LivenessChecker livenessChecker;
+  private final Map<Integer, KBucket> buckets = new HashMap<>();
+  private final Clock clock;
+
+  public KBuckets(
+      final Clock clock,
+      final LocalNodeRecordStore localNodeRecordStore,
+      final LivenessChecker livenessChecker) {
+    this.clock = clock;
+    this.localNodeRecordStore = localNodeRecordStore;
+    this.homeNodeId = localNodeRecordStore.getLocalNodeRecord().getNodeId();
+    this.livenessChecker = livenessChecker;
+  }
+
+  public synchronized Stream<NodeRecord> getLiveNodeRecords(int distance) {
+    if (distance == 0) {
+      return Stream.of(localNodeRecordStore.getLocalNodeRecord());
+    }
+    return getBucket(distance).stream().flatMap(bucket -> bucket.getLiveNodes().stream());
+  }
+
+  public synchronized Stream<NodeRecord> getAllNodeRecords(int distance) {
+    if (distance == 0) {
+      return Stream.of(localNodeRecordStore.getLocalNodeRecord());
+    }
+    return getBucket(distance).stream().flatMap(bucket -> bucket.getAllNodes().stream());
+  }
+
+  private Optional<KBucket> getBucket(final int distance) {
+    return Optional.ofNullable(buckets.get(distance));
+  }
+
+  public synchronized void offer(NodeRecord node) {
+    final int distance = Functions.logDistance(homeNodeId, node.getNodeId());
+    if (distance > MAXIMUM_BUCKET) {
+      // Distance too great, ignore.
+      return;
+    }
```

Comment:
if it's outside the bucket max should it be even in our list? I'm not sure I fully understand why we might store a node that is > 256 distance...

Reply:
This is a bit of a quirk because of the way `NodeRecord` is implemented.  The distance is the log distance so is actually the length of the common prefix of the two node IDs.  Since NodeIDs are a hash they are always 32 bytes and thus the greatest possible distance is 256, but `NodeRecord` uses `Bytes` not `Bytes32` as the type and has a pluggable system for calculating the node ID so we don't have any guarantee that it won't be longer than 32 bytes. So we wind up needing this check.